tools/autograd/gen_variable_type.py

```diff
@@ -318,7 +318,8 @@
 
 FW_DERIVATIVE_DEFINED_GRAD_TEMPLATE = CodeTemplate("""\
 auto ${inp}_t_raw = toNonOptFwGrad(${inp});
-auto ${inp}_t = ${inp}_t_raw.defined() ? ${inp}_t_raw : at::zeros_like(toNonOptTensor(${inp}));
+auto ${inp}_tensor = toNonOptTensor(${inp});
+auto ${inp}_t = ${inp}_t_raw.defined() ? ${inp}_t_raw : at::_efficientzerotensor(${inp}_tensor.sizes(), ${inp}_tensor.options());
```

Comment:
Does this PR only use efficient zero tensors in forward-mode AD? Are you planning on using them for reverse-mode AD as well?

Reply:
> Does this PR only use efficient zero tensors in forward-mode AD? 

yes

> Are you planning on using them for reverse-mode AD as well? 

There aren't that many cases where this would be useful in reverse-mode AD since very few functions have multiple outputs. But, yeah possibly! I see no reason not to.

> Where else are we planning on using efficient zeros tensors in the future?

Alban and I have discussed exposing this in public API via kwarg arg in `torch.zeros` but we haven't decided anything yet.

Reply:
After reading through this PR: we're going to need to modify functorch to handle efficient zero tensors correctly. In particular, whenever we create a wrapper Tensor, it should receive the zeros dispatch key, otherwise it will not report `_is_zerotensor` correctly.

There are implications for this for tensor subclassing too. In addition to sizes/strides being metadata on a wrapper tensor subclass, the wrapper tensor subclass also probably wants to be able to say that it has the {conjugate, neg, efficient zeros} bit set. (cc @albanD does that sound right?)

If we agree on the above two items then I think that's all we need to say that "efficient zeros tensor works with functorch" and I can fix the problem mentioned in the first paragraph when this PR goes in.

Reply:
> There are implications for this for tensor subclassing too.

Conjugation and ZT info are not tensor metadata, so I would expect it to work fine for tensor subclasses, if we propagate all the dispatch keys correctly when making tensor subclasses. So, then, this seems like a big limitation of `_make_wrapper_subclass` since it makes tensor subclasses non-composable with any dispatch key feature. 

I was wondering what if we instead add `_make_wrapper_subclass_like` which will have `_like` API semantics, and will allow us to propagate the original tensor's dispatch keys and set them on the wrapper tensor as well, with the option to override the tensor metadata through keyword arguments.

Reply:
If we go the `_make_wrapper_subclass_like` we should probably stop recommending people use `_make_wrapper_subclass` because that's a good way to shoot oneself in the foot